src/turbomind/layers/sampling_layers/LogitsProcessorLayer.cc

```diff
@@ -14,213 +14,226 @@
-                                   end_ids.shape[1],
+    if (end_ids_ten_) {
+        TM_CHECK_EQ(end_ids_ten_.ndim(), 2);
+        auto enable = [&] {
+            const int num_generated_tokens = step - max_input_length;
+            auto      prompt_len           = args.at("prompt_length").data<int>();
+            auto      context_len          = args.at("context_length").data<int>();
+            for (int i = 0; i < bsz; ++i) {
+                if (min_lengths_[i] > context_len[i] - prompt_len[i] + num_generated_tokens) {
+                    return true;
+                }
+            }
+            return false;
+        };
+        if (enable()) {
+            invokeMinLengthPenalty(logits.data(),
+                                   min_lengths_buf_.data(),
+                                   args.at("sequence_length").data<int>(),
+                                   vocab_size_padded_,
+                                   bsz,
+                                   end_ids_ten_.data(),
+                                   end_ids_ten_.shape(1),
                                    stream_);
             sync_check_cuda_error();
         }
     }
 
     // temperature
-    {
-        if (!ALL_OF(temperature_.begin(), batch_size, float, 1.f)) {
-            invokeBatchApplyTemperaturePenalty_v2(
-                logits, (T*)nullptr, temperature_buf_, batch_size, args_.vocab_size, args_.vocab_size_padded, stream_);
-            sync_check_cuda_error();
-        }
+    if (!ALL_OF(temperature_.begin(), bsz, float, 1.f)) {
+        invokeBatchApplyTemperaturePenalty_v2(logits.data(),  //
+                                              (T*)nullptr,
+                                              temperature_buf_.data(),
+                                              bsz,
+                                              vocab_size_,
+                                              vocab_size_padded_,
+                                              stream_);
+        sync_check_cuda_error();
     }
 
     TM_LOG_DEBUG("%s stop", __PRETTY_FUNCTION__);
 }
 
 template<typename T>
-void LogitsProcessorLayer<T>::setup(const size_t batch_size, const size_t beam_width, TensorMap* runtime_args)
+void LogitsProcessorLayer<T>::Setup(const std::vector<const Request*>& rs)
 {
     TM_LOG_DEBUG("%s start", __PRETTY_FUNCTION__);
 
-    allocateBuffer(batch_size);
+    const int bsz = rs.size();
 
-    // repetition_penalty
-    if (runtime_args->isExist("repetition_penalty")) {
-        init_host_buffer(runtime_args, "repetition_penalty", batch_size, repetition_penalty_.data(), 1.f);
-        repetition_penalty_type_ = RepetitionPenaltyType::Multiplicative;
+    for (int i = 0; i < bsz; ++i) {
+        auto& c = rs[i]->gen_cfg;
+        // repetition_penalty
+        repetition_penalty_[i] = c.repetition_penalty;
+        // temperature
+        temperature_[i] = c.temperature;
+        // min_length
+        min_lengths_[i] = c.min_new_tokens;
```

Comment:
the setup for `invokeMinLengthPenalty` is not right. `context_length_` and `prompt_length_` are not used.